aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,680 @@
+    auto pool_offsets_ptr = pool_offsets.data_ptr<int64_t>();
+
+    thrust::for_each(
+        policy,
+        thrust::make_counting_iterator(int64_t(0)),
+        thrust::make_counting_iterator(int64_t(new_sz)),
+        [values_accessor,
+         sorted_indices_ptr,
+         pool_sizes_ptr,
+         pool_offsets_ptr,
+         mx_buffer_ptr,
+         nvalues] __device__(int64_t index) {
+          int64_t curr_pool_size = pool_sizes_ptr[index];
+          auto mx_row = mx_buffer_ptr + index * nvalues;
+          int64_t offset = pool_offsets_ptr[index];
+          for (int64_t p = 0; p < curr_pool_size; p++) {
+            int64_t i = *(sorted_indices_ptr + offset + p);
+            auto values_row = values_accessor[i].data();
+            for (int64_t j = 0; j < nvalues; j++) {
+              mx_row[j] = c10::cuda::compat::max(mx_row[j], values_row[j]);
+            }
+          }
+        });
+  }
+  return std::make_tuple(
+      sorted_indices, pool_offsets, pool_sizes, mx_buffer);
+}
+
+template <typename scalar_t, bool LogSoftMax>
+void cuda_sparse_coo_softmax(
+    Tensor& output,
+    const Tensor& input,
+    const int64_t dim) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
+    implementation of the sparse softmax algorithm that this implementation is
+    based on.
+  */
+  auto sparse_dim = input.sparse_dim();
+  auto indices = input._indices().contiguous();
+  auto values = input._values().contiguous();
+  auto out_values = output._values();
+  auto out_indices = output._indices();
+  out_values.resize_as_(values);
+  out_indices.resize_as_(indices);
+  out_indices.copy_(indices);
+
+  if (dim >= sparse_dim) {
+    if (LogSoftMax) {
+      auto new_values = log_softmax_cuda(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    } else {
+      auto new_values = softmax_cuda(values, dim - sparse_dim + 1, false);
+      out_values.copy_(new_values);
+    }
+    return;
+  }
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  auto nnz = values.size(0);
+  auto sizes = input.sizes();
+  auto nvalues = get_nvalues(sizes, sparse_dim);
+
+  /* Prepare accessors */
+  auto values_2 = values.view({nnz, nvalues});
+  auto values_accessor = values_2.packed_accessor<scalar_t, 2>();
+
+  auto out_values_2 = out_values.view({nnz, nvalues});
+  auto out_values_accessor = out_values_2.packed_accessor<scalar_t, 2>();
+
+  Tensor sorted_indices;
+  Tensor pool_offsets;
+  Tensor pool_sizes;
+  Tensor mx_buffer;
+
+  std::tie(sorted_indices, pool_offsets, pool_sizes, mx_buffer) =
+      compute_pool_max<scalar_t, true>(indices, values_2, sizes, nvalues, dim);
+
+  auto pool_size = pool_offsets.size(0);
+  auto exp_sums_rows = at::empty({nvalues * pool_size}, values.options());
```

Comment:
Instead of filling can this tensor just be created with at::zeros? Or, even better, if these values are only written to and not read does this need to be filled at all?

Reply:
Yes, at::zeros is better, and as this buffer data is used as an accumulator inside the kernel, the initialization with zero is needed.